### ISSUE TITLE
Increased memory requirements for native-image compilation with Quarkus master

### ISSUE THREAD

**Issue by rsvoboda** [quarkusio/quarkus#15706]
We see increased memory requirements in BeefyTS  (module 003-quarkus-many-extensions) for native-image compilation with Quarkus master comparing to Quarkus 1.12.2.Final

Run with Quarkus 1.12.2.Final finishes successfully, run with Quarkus master finishes with `java.lang.OutOfMemoryError: GC overhead limit exceeded`. We use `-Dquarkus.native.native-image-xmx=4g` in our GitHub Action.

I was able to identify that quarkus-kubernetes / quarkus-openshift extension pulls in additional 10+ MB of dependencies - filed https://github.com/quarkusio/quarkus/issues/15705. This is affecting native-image execution because more classes need to be analyzed and thus increases pressure on heap space.

The problem is not only quarkus-kubernetes / quarkus-openshift extension. When I removed `quarkus-kubernetes` extension from 003-quarkus-many-extensions module I was able to get the native binary with `-Dquarkus.native.native-image-xmx=4g` setting.
But the build takes longer comparing to 1.12.2.Final, 08:47 min for Quarkus master versus 07:41 min for Quarkus 1.12.2.Final.
This is something to be tackled on Quarkus side to identify other extensions which are causing increased memory requirements.

Investigation of native-image related issues is really problematic, takes time and overloads the machine for several minutes. I tried few native builds with different sets of extensions to see how 1.12.2.Final and Quarkus master differs. In the end I went back to JVM mode and compared pulled jars to identify quarkus-kubernetes issue.
I miss something handy which would provide more details when comparing two builds of the native-image.
This is something to be tackled preferable by Mandrel team, or maybe escalated by them to upstream for wider discussion. 
@galderz fyi^

Reproducer:
```
git clone git@github.com:quarkus-qe/beefy-scenarios.git
cd beefy-scenarios && git checkout e83bcc8

mvn -s .github/mvn-settings.xml clean package -DskipTests -Dnative \
    -Dquarkus.native.container-build=true -Dquarkus.native.native-image-xmx=4g \
    -Dquarkus.platform.version=1.12.2.Final -Dquarkus-plugin.version=1.12.2.Final  \
    -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-native-image:21.0-java11 \
    -pl '003-quarkus-many-extensions'

mvn -s .github/mvn-settings.xml clean package -DskipTests -Dnative \
    -Dquarkus.native.container-build=true -Dquarkus.native.native-image-xmx=4g \
    -Dquarkus.platform.version=999-SNAPSHOT -Dquarkus-plugin.version=999-SNAPSHOT  \
    -Dquarkus.native.builder-image=quay.io/quarkus/ubi-quarkus-native-image:21.0-java11 \
    -pl '003-quarkus-many-extensions'
```

**Comment by famod** [comment]
Maybe this also explains that: https://github.com/quarkusio/quarkus/pull/15698#issuecomment-798622287

**Comment by zakkak** [comment]
> I miss something handy which would provide more details when comparing two builds of the native-image.

@rsvoboda Could you please give a list of the info/details you would be interested in?

**Comment by rsvoboda** [comment]
My complain is about the pain with native while trying to investigate what changed between two versions of Quarkus.

Basically something that could gather details from native-image build with Quarkus version and A and later with Quarkus version B and compare that gathered data to tell me that the run with B used extra 2 GB of memory more because these and these data structures took extra 1 GB of data, XYZ took 500 MB of data and ABC together with CDF took the remaining 500 MB of the additional memory.

GraalVM / native-image feels like we're going back productivity wise, we need to improve this experience. For both users/customers and also for us. It's such a pitta to do anything with native and quite often waste of a lot of hours. Plus few tears I drop for my machine when overheating the CPU and spinning the cooler to max speed.

In short, slow, clunky, go-get-a-coffee experience which didn't change in last 2-3 years, when native-image was getting some traction. I would even say that old GraalVM 1.0.0-rcXYZ builds were faster, but maybe it's just me and my selective memories.

**Comment by galderz** [comment]
Re: I miss something handy which would provide more details when comparing two builds of the native-image.

[`-Dquarkus.native.enable-reports`](https://quarkus.io/guides/building-native-image#quarkus-native-pkg-native-config_quarkus.native.enable-reports) is your friend. It outputs several files, including:

* A call tree (useful for seeing which path leads to a certain method being included)

* List of classes and methods included in call tree (useful for comparing with other versions.

**Comment by galderz** [comment]
> I would even say that old GraalVM 1.0.0-rcXYZ builds were faster, but maybe it's just me and my selective memories.

See the comment I posted earlier today [here](https://github.com/oracle/graal/issues/2129#issuecomment-804109497). The more classes/methods that are part of the application, the longer it takes.

**Comment by geoand** [comment]
Is this still an issue or can we close it?

**Comment by zakkak** [comment]
@rsvoboda FYI with the new native-image output it's even easier to find major differences between two compilations. See https://github.com/oracle/graal/blob/master/docs/reference-manual/native-image/BuildOutput.md

**Comment by geoand** [comment]
Closing as `out of date`. If it's still an issue, feel free to reopen
